internal/nix/search.go

```diff
@@ -106,3 +111,78 @@ func searchSystem(url, system string) (map[string]*Info, error) {
 	}
 	return parseSearchResults(out), nil
 }
+
+type CachedSearchResult struct {
+	Results map[string]*Info `json:"results"`
+	// Query is added easier to grep for debuggability
+	Query string `json:"query"`
+}
+
+// SearchNixpkgsAttribute is a wrapper around searchSystem that caches results.
+// NOTE: we should be very conservative in where we use this function. `nix search`
+// accepts generalized `installable regex` as arguments but is slow. For certain
+// queries of the form `nixpkgs/<commit-hash>#attribute`, we can know for sure that
+// once `nix search` returns a valid result, it will always be the very same result.
+// Hence we can cache it locally and answer future queries fast, by not calling `nix search`.
+func SearchNixpkgsAttribute(query string) (map[string]*Info, error) {
+	key := cacheKey(query)
+
+	// Check if the query was already cached, and return the result if so
+	cache := filecache.New("devbox/nix", filecache.WithCacheDir(xdg.CacheSubpath("")))
+	if cachedResults, err := cache.Get(key); err == nil {
+		var results map[string]*Info
+		if err := json.Unmarshal(cachedResults, &results); err != nil {
+			return nil, err
+		}
+		return results, nil
+	} else if !filecacheNeedsUpdate(err) {
+		return nil, err // genuine error
+	}
+
+	// If not cached, or an update is needed, then call searchSystem
+	infos, err := searchSystem(query, "" /*system*/)
+	if err != nil {
+		return nil, err
+	}
+
+	// Save the results to the cache
+	marshalled, err := json.Marshal(infos)
+	if err != nil {
+		return nil, err
+	}
+	// TODO savil: add a SetForever API that does not expire. Time based expiration is not needed here
+	// because we're caching results that are guaranteed to be stable.
+	// TODO savil: Make filecache.cache a public struct so it can be passed into other functions
+	const oneYear = 12 * 30 * 24 * time.Hour
+	if err := cache.Set(key, marshalled, oneYear); err != nil {
+		return nil, err
+	}
+
+	return infos, nil
+}
+
+// read as: filecache.NeedsUpdate(err)
+// TODO savil: this should be implemented in the filecache package
+func filecacheNeedsUpdate(err error) bool {
+	return errors.Is(err, filecache.NotFound) || errors.Is(err, filecache.Expired)
+}
+
+// cacheKey sanitizes the search query to be a valid unix filename.
+// This cache key is used as the filename to store the cache value, and having a
+// representation of the query is important for debuggability.
+func cacheKey(input string) string {
```

Comment:
Do we handle collisions for queries that have sanitized characters? Using the infamous `emacsPackages.@` package as an example, the input would get cached under the key `emacsPackages.`. Does that mean `devbox add emacsPackages.` would install that package if it was in the cache?

Reply:
Ooh good catch. I added that `@` to the "replace disallowed chars with underscores"

I tried searching the other non-alphanumeric-characters on my standard American keyboard, and no packages showed up so I think its okay (practically speaking). 

Theoretically speaking this scheme is not collision-resistant since two packages with `foo-<disallowed-char-number-one>` and `foo-<disallowed-char-number-two>` would have a collision in the sanitized name. Same for the name-length-truncation we do.